File: src/main/java/Decorator/Observer.java
package Decorator;

import java.util.ArrayList;

interface IObserver {
    void update();
}

class Observer implements IObserver {
    @Override
    public void update() {
        System.out.println("observer got notified by Observable");
    }
}

class Observable {
    private ArrayList<IObserver> observers = new ArrayList<>();

    public void register(IObserver observer) {
        observers.add(observer);
    }

    public void unRegister(IObserver observer) {
        observers.remove(observer);
    }

    public void onEvent() {
        notifyAllObservers();
        System.out.println("doing event");
    }

    private void notifyAllObservers() {
        for (IObserver observer : observers) {
            observer.update();
        }
    }
}

File: src/main/kotlin/Observer.kt
class Observable {
    private var observables = ArrayList<IObserver>()
    fun register(observer: IObserver) = observables.add(observer)
    fun unRegister(observer: IObserver) = observables.removeIf { it == observer }

    fun onEvent() {
        print("event handled")
        notifyObservers()
    }

    private fun notifyObservers() = observables.forEach { it.update() }
}


interface IObserver {
    fun update()
}

class Observer : IObserver {
    override fun update() = print("Observer got updated")
}